test/upgrade/prober/verify.go

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2020 The Knative Authors
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package prober
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/wavesoftware/go-ensure"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"knative.dev/eventing/test/lib/nodes"
+)
+
+func (p *prober) Verify() ([]error, int) {
+	nc := nodes.Client(p.client.Kube, p.log)
+	node, err := nc.RandomWorkerNode()
+	ensure.NoError(err)
+	address := nc.GuessNodeExternalAddress(node)
+	p.log.Debugf("Address resolved: %v, type: %v", address.Address, address.Type)
+	report := fetchReceiverReport(address, p.log)
+	p.log.Infof("Fetched receiver report. Events propagated: %v. "+
+		"State: %v", report.Events, report.State)
+	if report.State == "active" {
+		panic(errors.New("report fetched to early, receiver is in active state"))
+	}
+	errs := make([]error, 0)
+	for _, t := range report.Thrown {
+		if strings.Contains(t, "should be received once") {
+			p.log.Warn("Duplicate events: ", t)
```

Comment:
Should this be an error? Not sure what the purpose of the warning or what someone who sees it should do?

Reply:
Duplicate events are not really errors as our guarantee is "at least once". So we just log it here. If later we see this is a concerning problem, we can make it error.